### DISCUSSION TITLE
Minimum C Version

### DISCUSSION THREAD

**Discussion by Yay295** [python-pillow/Pillow#6516] [General]
While looking through the C code I noticed that loop variables are always declared before the loop
```c
int x;
for (x = 0; x < size; ++x) {
```
instead of in it
```c
for (int x = 0; x < size; ++x) {
```
except in /src/thirdparty/raqm/raqm.c and in this one location https://github.com/python-pillow/Pillow/blob/4c59f77e37e7333544a5d0910de8148f7972ed06/src/display.c#L519

C added support for this in C99, and a lot of this code is older than that so it makes sense that it wasn't used. But 1999 was also 23 years ago now. Should we still be trying to code to pre-C99, or should we start using a more modern standard? The makefile doesn't specify a standard to use.

**Comment by nulano** [comment]
Re raqm, that is a third party library included as-is (except for a small fribidi header included patch), and only used if you specify a specific build flag when installing Pillow. When I tried making that the default, I found a lot of the Linux systems tested in CI use a GCC version that defaults to C89 standard causing several compile errors in raqm. The line you found in display.c is also Windows-only, so will never be seen by GCC, only MSVC which now supports C99 by default (but didn't in the past).

> But 1999 was also 23 years ago

I would say take that up with the GCC devs, but they have finally made the default a newer standard somewhat recently.

**Reply by hugovk** [reply]
We don't have a minimum C version as such, other than [this](https://pillow.readthedocs.io/en/stable/installation/platform-support.html#platform-support):

> Binary distributions are contributed for each release on a volunteer basis, but the source should compile and run everywhere platform support is listed. In general, we aim to support all current versions of Linux, macOS, and Windows.

---

A quick test after changing a couple of them, all CI jobs passed except for CentOS 7:

```
      src/libImaging/Convert.c: In function 'bit2l':
      src/libImaging/Convert.c:59:5: error: 'for' loop initial declarations are only allowed in C99 mode
           for (int x = 0; x < xsize; x++) *out++ = (*in++ != 0) ? 255 : 0;
           ^
      src/libImaging/Convert.c:59:5: note: use option -std=c99 or -std=gnu99 to compile your code
```

https://github.com/hugovk/Pillow/actions/runs/2902666901

CentOS 7 is [still supported, until EOL on 30 Jun 2024](https://endoflife.date/centos).

**Comment by Yay295** [comment]
I found these two lines:
https://github.com/python-pillow/Pillow/blob/196210bc804a4575b6cc0b1cb6c9b7b1f09e4ff9/src/libImaging/Draw.c#L38
https://github.com/python-pillow/Pillow/blob/196210bc804a4575b6cc0b1cb6c9b7b1f09e4ff9/src/libImaging/QuantTypes.h#L19

`stdint.h` is technically a C99 addition, though it could have been provided earlier. Also both of those files can be disabled with defines (`WITH_IMAGEDRAW` and `WITH_QUANTIZE`). But if this include is working on all supported platforms, we could simplify this section:
https://github.com/python-pillow/Pillow/blob/196210bc804a4575b6cc0b1cb6c9b7b1f09e4ff9/src/libImaging/ImPlatform.h#L41-L71

**Comment by raygard** [comment]
Still not clear to me if C99 should be considered minimum version for building Pillow. As I mentioned in https://github.com/python-pillow/Pillow/pull/5367#issuecomment-811159299 I believe these lines in Quant.c uses a compound literal, which came in C99:

```
            dwi[j] = (DistanceWithIndex){
                &(avgDist[i * nEntries + j]),
                j
            };
```

Correct me if I am mistaken. Maybe GCC supports this without specifying C99? Are there other places that use compound literals?

**Reply by Yay295** [reply]
> Maybe GCC supports this without specifying C99?

Yes https://gcc.gnu.org/onlinedocs/gcc/C-Extensions.html

Also https://gcc.gnu.org/onlinedocs/gcc/Standards.html

I believe the default standard used by GCC on CentOS 7 is `gnu90`.

**Comment by radarhere** [comment]
#8644 has now removed support for pre-C99.
